src/runtime-rs/crates/kata/src/task_service.rs

```diff
@@ -0,0 +1,874 @@
+                .send(msg)
+                .map_err(|e| {
+                    get_rpc_status(
+                        Code::INTERNAL,
+                        format!("Failed to shutdown with error: {:?}", e),
+                    )
+                })?;
+        } else {
+            debug!(
+                sl!(),
+                "<==== shutdown: cannot shutdown shim, containers not empty"
+            );
+        }
+
+        Ok(Empty::new())
+    }
+}
+
+impl TaskService {
+    fn do_state(&self, req: shim::StateRequest) -> Result<shim::StateResponse> {
+        let cid = req.get_id();
+        let eid = req.get_exec_id();
+
+        let mut sandbox = self.sandbox.lock().expect("poisoned sandbox lock");
+        let c = sandbox.find_container(cid).map_err(|e| {
+            get_rpc_status(
+                Code::INTERNAL,
+                format!("failed to find container {}  error {}", cid, e),
+            )
+        })?;
+        // unlock the sandbox
+        drop(sandbox);
+
+        let container = c.lock().expect("poisoned container lock");
+        let bundle = container.config.bundle_path.clone();
+        if eid.is_empty() {
+            let status = to_status(container.common_process.state);
+
+            let (exit_code, exit_time) = exit_info(container.common_process.status.clone());
+
+            let stdin = container.common_process.stdin.clone().unwrap_or_default();
+            let stdout = container.common_process.stdout.clone().unwrap_or_default();
+            let stderr = container.common_process.stderr.clone().unwrap_or_default();
+            let terminal = container.common_process.terminal;
+            Ok(shim::StateResponse {
+                id: cid.into(),
+                bundle,
+                pid: self.pid,
+                status,
+                stdin,
+                stdout,
+                stderr,
+                terminal,
+                exit_status: exit_code,
+                exited_at: exit_time.into(),
+                ..Default::default()
+            })
+        } else if let Some(exec) = container.processes.get(eid) {
+            let status = to_status(exec.state);
+            let (exit_code, exit_time) = exit_info(exec.common_process.status.clone());
+            let stdin = exec.common_process.stdin.clone().unwrap_or_default();
+            let stdout = exec.common_process.stdout.clone().unwrap_or_default();
+            let stderr = exec.common_process.stderr.clone().unwrap_or_default();
+            let terminal = exec.common_process.terminal;
+            Ok(shim::StateResponse {
+                id: cid.into(),
+                bundle,
+                pid: self.pid,
+                status,
+                stdin,
+                stdout,
+                stderr,
+                terminal,
+                exit_status: exit_code,
+                exited_at: exit_time.into(),
+                ..Default::default()
+            })
+        } else {
```

Comment:
There is a lot of commonality in these two blocks so I wonder if this could be simplified by calling a helper method?

Reply:
I agree. Pulling out the generic functionality here would make this a lot easier to follow, and would prevent some duplication.